File: test/native/message_test.cpp
// Native message delivery test for the interactive bridge.
//
// The assignment problem pairs a client that sends a request with a server that waits to receive one.
// The message is not explicitly addressed, so its delivery is not resolved automatically but surfaced
// to the caller as a message delivery decision. This drives one client and one server and checks that
// the engine stops at the delivery, offering the waiting message identified by its origin and sender
// from the header, that enqueuing that identity delivers the message, and that both the sending and
// the receiving task then complete. The message content is derived by the engine from status and data
// and plays no part in the identity the bridge uses.

#include <cstdlib>
#include <fstream>
#include <iostream>
#include <memory>
#include <sstream>
#include <string>
#include <vector>

#include "Controller.h"
#include "Engine.h"
#include "Input.h"
#include "Monitor.h"

using namespace BPMNOS::WASM;
using namespace BPMNOS;
using json = nlohmann::ordered_json;

static std::string readFile(const std::string& path) {
  std::ifstream file(path, std::ios::binary);
  if (!file) {
    std::cerr << "cannot open " << path << "\n";
    std::exit(2);
  }
  std::stringstream buffer;
  buffer << file.rdbuf();
  return buffer.str();
}

static void check(bool condition, const std::string& message) {
  if (!condition) {
    std::cerr << "FAIL: " << message << "\n";
    std::exit(1);
  }
  std::cerr << "ok: " << message << "\n";
}

int main(int argc, char** argv) {
  std::string fixtureDir = (argc > 1) ? argv[1] : "test/fixtures";
  std::string modelXml = readFile(fixtureDir + "/Assignment_problem.bpmn");
  std::string costsCsv = readFile(fixtureDir + "/costs.csv");
  std::string instanceCsv =
    "INSTANCE_ID; NODE_ID; INITIALIZATION\n"
    "Client1; ClientProcess;\n"
    "Server1; ServerProcess;\n";

  Input input(modelXml);
  json required = input.requiredLookupTables();
  check(required.is_array() && required.size() == 1 && required[0] == "costs.csv",
        "requiredLookupTables reports the model's lookup table");
  input.addLookupTable("costs.csv", costsCsv);
  input.setInstance(instanceCsv);
  Monitor monitor;
  Controller controller;
  Engine engine(input.release(), Engine::Config{}, &monitor, &controller);

  json log = json::array();
  monitor.addObserver([&](const json& entry) { log.push_back(entry); });

  engine.run();

  auto findMessageRequest = [&]() -> std::shared_ptr<const Execution::DecisionRequest> {
    for (const auto& weak : controller.getPendingRequests()) {
      auto request = weak.lock();
      if (request && request->type == Execution::Observable::Type::MessageDeliveryRequest) {
        return request;
      }
    }
    return nullptr;
  };

  auto request = findMessageRequest();
  check(request != nullptr, "the engine stopped at the message delivery");

  int delivered = 0;
  int guard = 0;
  while (request && guard++ < 50) {
    auto candidates = controller.getMessageCandidates(request.get());
    check(!candidates.empty(), "the delivery offers at least one candidate message");
    check(controller.enqueueMessageDeliveryDecision(request, candidates.front()).has_value(),
          "enqueueMessageDeliveryDecision accepted");
    ++delivered;
    engine.resume();
    request = findMessageRequest();
  }
  check(delivered == 1, "exactly one message was delivered");

  bool sent = false;
  bool received = false;
  for (const auto& entry : log) {
    if (!entry.contains("token")) {
      continue;
    }
    const auto& token = entry["token"];
    if (token.value("state", std::string()) != "COMPLETED") {
      continue;
    }
    std::string nodeId = token.value("nodeId", std::string());
    if (nodeId == "SendRequestTask") sent = true;
    else if (nodeId == "ReceiveRequestTask") received = true;
  }
  check(sent, "the send task completed");
  check(received, "the receive task completed");

  std::cerr << "ALL PASSED (message delivery)\n";
  return 0;
}

File: test/wasm/message_test.mjs
// WebAssembly message delivery test. The message of the assignment problem is not explicitly addressed,
// so its delivery is surfaced to the caller, who names the waiting message by its origin and sender from
// the header. This mirrors the native message test through the module's JavaScript interface.

import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { dirname, join } from 'node:path';
import createBPMNOS from '../../dist/bpmnos.mjs';
import { interactive } from './composition.mjs';

const here = dirname(fileURLToPath(import.meta.url));
const root = join(here, '..', '..');

function check(condition, message) {
  if (!condition) {
    console.error(`FAIL: ${message}`);
    process.exit(1);
  }
  console.error(`ok: ${message}`);
}

const module = await createBPMNOS();

const modelXml = readFileSync(join(root, 'test', 'fixtures', 'Assignment_problem.bpmn'), 'utf8');
const costsCsv = readFileSync(join(root, 'test', 'fixtures', 'costs.csv'), 'utf8');
const instanceCsv =
  'INSTANCE_ID; NODE_ID; INITIALIZATION\n' +
  'Client1; ClientProcess;\n' +
  'Server1; ServerProcess;\n';

const input = new module.Input(modelXml);
const required = JSON.parse(input.getLookupTableNames());
check(Array.isArray(required) && required.length === 1 && required[0] === 'costs.csv',
  "getLookupTableNames reports the model's lookup table");
input.addLookupTable('costs.csv', costsCsv);
input.setInstance(instanceCsv);
const monitor = new module.Monitor();
const controller = new module.Controller(interactive);
const engine = new module.Engine(input, JSON.stringify({ provider: 'static' }), controller, monitor);
input.delete();

const log = [];
monitor.addObserver((entryJson) => log.push(JSON.parse(entryJson)));

engine.run(0);
let pending = JSON.parse(controller.getPendingDecisions());
check(pending.length > 0, 'the engine stopped at the message delivery');

// What a caller replaying the stream does: the messages it has seen created and not yet seen disposed of,
// matched against the criterion the delivery request carried. The engine is not asked, and needs not be.
function heldMessages() {
  const held = new Map();

  for (const entry of log) {
    if (!entry.message) {
      continue;
    }
    const key = `${entry.message.origin}|${entry.message.header.sender}`;

    if (entry.message.state === 'CREATED') {
      held.set(key, entry.message);
    } else {
      held.delete(key);
    }
  }

  return [ ...held.values() ];
}

function matches(message, criterion) {
  const expected = criterion.recipientHeader;

  return criterion.senders.includes(message.origin)
    && Object.keys(expected).length === Object.keys(message.header).length
    && Object.entries(expected).every(([ key, value ]) =>
      value === null || message.header[key] === null || message.header[key] === value);
}

let delivered = 0;
let guard = 0;
while (pending.length > 0 && guard++ < 50) {
  const request = pending[0];
  check(request.type === 'messageDelivery', 'the pending decision is a message delivery');

  const criterion = log
    .filter((entry) => entry.messageDeliveryRequest
      && entry.messageDeliveryRequest.instanceId === request.instanceId
      && entry.messageDeliveryRequest.nodeId === request.nodeId)
    .pop();

  check(criterion !== undefined, 'the request was announced as a record');
  check(Array.isArray(criterion.messageDeliveryRequest.senders),
    'carrying the senders the token accepts');
  check(typeof criterion.messageDeliveryRequest.recipientHeader === 'object',
    'and the header it expects');

  const candidates = heldMessages().filter((message) =>
    matches(message, criterion.messageDeliveryRequest));

  check(candidates.length > 0, 'a message held matches what the token waits for');
  const candidate = candidates[0];
  const decision = {
    instanceId: request.instanceId,
    nodeId: request.nodeId,
    origin: candidate.origin,
    sender: candidate.header.sender,
  };
  check(!('rejected' in JSON.parse(controller.enqueueMessageDeliveryDecision(JSON.stringify(decision)))),
    'enqueueMessageDeliveryDecision accepted');
  delivered += 1;
  engine.resume();
  pending = JSON.parse(controller.getPendingDecisions());
}

check(delivered === 1, 'exactly one message was delivered');

const completed = new Set(
  log.filter((e) => e.token && e.token.state === 'COMPLETED').map((e) => e.token.nodeId));
check(completed.has('SendRequestTask'), 'the send task completed');
check(completed.has('ReceiveRequestTask'), 'the receive task completed');

console.error('ALL PASSED (WebAssembly message delivery)');
